### DISCUSSION TITLE
NextJS API routes very large on Vercel, causing very slow cold start

### DISCUSSION THREAD

**Discussion by TheUnlocked** [vercel/community#836] [Help] [answered]
For reference, the codebase is at https://github.com/TheUnlocked/Necode/tree/main

On Vercel, if I set `NEXT_DEBUG_FUNCTION_SIZE=1`, I see some strange dependencies for my API routes:

```
Serverless Function's pages: api/auth/[...nextauth].js, api/classroom.js, api/classroom/[classroomId].js, api/classroom/[classroomId]/activity/[activityId].js, api/classroom/[classroomId]/activity/ice.js, api/classroom/[classroomId]/activity/live.js, api/classroom/[classroomId]/activity/submission.js, api/classroom/[classroomId]/join-code.js, api/classroom/[classroomId]/lesson.js, api/classroom/[classroomId]/lesson/[lessonId].js, api/classroom/[classroomId]/lesson/[lessonId]/activity.js, api/classroom/[classroomId]/me.js, api/classroom/[classroomId]/members.js, api/classroom/[classroomId]/members/[userId].js, api/classroom/join.js, api/me.js, api/users.js, api/users/[userId].js, api/users/simulated.js
--
00:36:54.490 | Large Dependencies                Uncompressed size  Compressed size
00:36:54.490 | node_modules/.prisma/client                 44.6 MB          15.6 MB
00:36:54.490 | node_modules/@mui/icons-material            10.8 MB          4.72 MB
00:36:54.491 | node_modules/next/dist                      4.24 MB          1.28 MB
00:36:54.491 | node_modules/@mui/material                  1.95 MB           596 kB
00:36:54.491 | node_modules/react-dom/cjs                  2.28 MB           558 kB
00:36:54.491 | node_modules/@prisma/client                 1.19 MB           260 kB
00:36:54.491 | node_modules/@mui/x-date-pickers             549 kB           166 kB
00:36:54.491 | node_modules/@mui/base                       486 kB           141 kB
00:36:54.491 | node_modules/lodash/lodash.js                544 kB          96.4 kB
00:36:54.491 |  
00:36:54.491 | All dependencies                            71.5 MB          24.6 MB
```

For some reason frontend code (I suspect the frontend code associated with `_app.ts`, though I'm not entirely sure) is included in the lambdas, a lot of it doesn't seem to have undergone much if any tree shaking, and the resulting output is massive. I would try to inspect the actual output files to get more details, but from what I can tell Vercel doesn't seem to let me do that.

This all results in very slow cold starts (~10 seconds), probably because loading 71MB of Javascript takes a bit of time.

What is causing this, and is there anything I can do to fix it?

**Comment by TheUnlocked** [comment]
Confirmed that it is including `_app.tsx`. After removing it for testing, the vercel build shows this:

```
Serverless Function's pages: api/auth/[...nextauth].js, api/classroom.js, api/classroom/[classroomId].js, api/classroom/[classroomId]/activity/[activityId].js, api/classroom/[classroomId]/activity/ice.js, api/classroom/[classroomId]/activity/live.js, api/classroom/[classroomId]/activity/submission.js, api/classroom/[classroomId]/join-code.js, api/classroom/[classroomId]/lesson.js, api/classroom/[classroomId]/lesson/[lessonId].js, api/classroom/[classroomId]/lesson/[lessonId]/activity.js, api/classroom/[classroomId]/me.js, api/classroom/[classroomId]/members.js, api/classroom/[classroomId]/members/[userId].js, api/classroom/join.js, api/me.js, api/users.js, api/users/[userId].js, api/users/simulated.js
--
11:08:30.964 | Large Dependencies             Uncompressed size  Compressed size
11:08:30.964 | node_modules/.prisma/client              44.6 MB          15.6 MB
11:08:30.964 | node_modules/next/dist                   4.11 MB          1.24 MB
11:08:30.965 | node_modules/react-dom/cjs               1.12 MB           286 kB
11:08:30.965 | node_modules/@prisma/client              1.19 MB           260 kB
11:08:30.965 | node_modules/lodash/lodash.js             544 kB          96.4 kB
11:08:30.965 |  
11:08:30.965 | All dependencies                         53.6 MB          18.1 MB
```

Why is `_app.tsx` being bundled with serverless function lambdas?

**Comment by amyegan** [comment, marked as answer]
Hi @TheUnlocked! Serverless Functions are bundled together for Next.js API routes until they are too big. This allows Next.js apps to bypass the [Serverless Function count limit](https://vercel.com/docs/concepts/limits/overview#serverless-functions-created-per-deployment) of 12 on Hobby accounts

You can explicitly define Serverless Functions via the [Function property in a `vercel.json` Project config](https://vercel.com/docs/project-configuration#project-configuration/functions). Using the Function property would allow you to move away from the default behavior described above and have separate Serverless Functions (i.e. not include the _app).

It may be a good idea to check your region to help improve speed as well. There is a guide on [improving Lambda cold start performance](https://vercel.com/guides/how-can-i-improve-serverless-function-lambda-cold-start-performance-on-vercel#improving-your-lambda-performance) that could help you identify other possible areas to modify.

**Reply by TheUnlocked** [reply]
@amyegan Thanks for replying. I'm fine with them being bundled (I prefer it actually, since it allows warming up one route to warm up the other routes). What I don't get is why `_app` would be included in the bundle when none of the routes listed for that lambda reference the frontend (they're all API routes). Fwiw, I've checked the dependencies for every file in `pages/api/**/*` using `@vercel/nft` and I don't see any reference to any frontend libraries like mui (which you can see is included in the lambda bundle on Vercel).

**Reply by amyegan** [reply]
I asked the Next.js team for a better explanation that what I could give. Here's what they said:

> For non-API specific serverless functions _app is needed for server-rendering the same as _document
>
>For API specific we currently include it as we have a pre-import step for _app and _document where setup logic is able to be handled 

They also said that it might not be necessary to include for all API cases and they're adding a task to their list to investigate that possibility. Thank you for bringing this forward!
